### PR TITLE
Implement scatter/gather for CubedSphereCommunicator

### PR DESCRIPTION
**Pull request by mcgibbon**
In preparing examples for publication, it will be helpful to plot global variables. This PR adds a cube gather routine to enable this. Scatter is added as well for completeness and other use cases, and because it was simple to do so. Most of the implementation is a re-use of the code for tile scatter and gather.

The current implementation requires (and enforces with NotImplementedError) that the tile dimension be the first dimension of the global Quantity. This may or may not be relaxed in the future. The rest of the dimension order is unrestricted.

Changes:
- Added TILE_DIM constant for tile dimension of global quantities
- Added Partitioner base class implementing features necessary for scatter/gather
- Moved scatter and gather from TileCommunicator to the Communicator base class, so its code can be re-used by the CubedSphereCommunicator
- implemented subtile_slice, global_extent, and subtile_extent routines on CubedSpherePartitioner necessary for scatter/gather in CubedSphereCommunicator.
- renamed argument `tile_extent` and `tile_dims` to `global_extent` and `global_dims` in routines to refer generically to the tile in the case of tile scatter/gather or cube in the case of cube scatter/gather

Tests are added for scatter/gather on the cube. While direct tests are not included for CubeSpherePartitioner's subtile_slice, global_extent, or subtile_extent routines, these are tested by the scatter/gather tests, which only assume TilePartitioner's versions of these routines are correct.

### CODE REVIEW

**Review by spencerkclark — APPROVED** [review]
This is cool!  It generally looks nice to me.  Just a few minor comments/questions.

File: HISTORY.md [hunk 1]
```diff
@@ -18,6 +18,11 @@ in the `coupler_nml` namelist.
 - Quantity.data is now guaranteed to be a numpy or cupy array matching its `.np` module, and will no longer be a gt4py Storage
 - Quantity accepts a `gt4py_backend` on initialize which is used to create its `.storage` if one was not used on initialize
 - parent MPI rank now referred to as "root" rank in variable names and documentation
+- Added TILE_DIM constant for tile dimension of global quantities
+- Added Partitioner base class implementing features necessary for scatter/gather
+- Moved scatter and gather from TileCommunicator to the Communicator base class, so its code can be re-used by the CubedSphereCommunicator
+- implemented subtile_slice, global_extent, and subtile_extent routines on CubedSpherePartitioner necessary for scatter/gather in CubedSphereCommunicator.
+- renamed argument `tile_extent` and `tile_dims` to `global_extent` and `global_dims` in routines to refer generically to the tile in the case of tile scatter/gather or cube in the case of cube scatter/gather
```
Comment (jdahm) [inline HISTORY.md:25]:
```suggestion
- Implemented subtile_slice, global_extent, and subtile_extent routines on CubedSpherePartitioner necessary for scatter/gather in CubedSphereCommunicator.
- Renamed argument `tile_extent` and `tile_dims` to `global_extent` and `global_dims` in routines to refer generically to the tile in the case of tile scatter/gather or cube in the case of cube scatter/gather.
```

**Review by jdahm — APPROVED** [review]

File: HISTORY.md (hunk 1 above)
Comment (mcgibbon) [reply HISTORY.md]: Capitalized but did not add periods, to be consistent with most of the other lines.

File: fv3gfs/util/partitioner.py [hunk 2]
```diff
@@ -499,6 +552,69 @@ def _get_corner(
             n_clockwise_rotations=rotations,
         )
 
+    def global_extent(self, rank_metadata: QuantityMetadata) -> Tuple[int, ...]:
+        """Return the shape of a full tile representation for the given dimensions.
```
[the remark this answers is not in the document]
Comment (mcgibbon) [reply fv3gfs/util/partitioner.py]: Thanks, swapped "tile" for "cube".

File: tests/test_cube_scatter_gather.py [hunk 3]
```diff
@@ -0,0 +1,341 @@
...
+
+
+@pytest.fixture(params=[0, 3])
+def n_tile_halo(request):
+    return request.param
+
+
+@pytest.fixture(params=["x,y", "y,x", "xi,y", "x,y,z", "z,y,x", "y,z,x"])
+def dims(request, fast):
+    if request.param == "x,y":
```
[the remark this answers is not in the document]
Comment (mcgibbon) [reply tests/test_cube_scatter_gather.py:29]: This suggestion is significantly more elegant, but it was harder for me to reason about what gets returned than what's currently here (which is more explicit). I'll leave this as is for now.

File: tests/test_cube_scatter_gather.py [hunk 4]
```diff
@@ -0,0 +1,341 @@
...
+            result = communicator.scatter_state(recv_state=state)
+        assert result["time"] == time
+        assert result["air_temperature"] is recv
+    for rank, (result, scattered) in enumerate(
+        zip(recv_quantities, scattered_quantities)
+    ):
+        assert result.dims == scattered.dims
+        assert result.units == scattered.units
+        assert result.extent == scattered.extent
+        scattered.np.testing.assert_array_equal(result.view[:], scattered.view[:])
```
[the remark this answers is not in the document]
Comment (mcgibbon) [reply tests/test_cube_scatter_gather.py]: Good suggestion, implemented. This ended up improving the tests by adding a check, that the type of result.data is correct, which was only implemented in one test.

**Pull request merged**